pkg/serviceaccounttoken/secret.go

```diff
@@ -44,10 +44,13 @@ func EnsureSecretForServiceAccount(ctx context.Context, secretsCache corecontrol
 	logrus.Tracef("EnsureSecretForServiceAccount for %s/%s", sa.GetNamespace(), sa.GetName())
 
 	// Lock avoids multiple calls to this func at the same time and creation of same resources multiple times
-	// Mutex is a addition to the Lease, it helps sync within the pod and avoid multiple Lease waits from the same pod
-	mutex := getLock(fmt.Sprintf("%v-%v", sa.Namespace, sa.Name))
+	lockKey := fmt.Sprintf("%v-%v", sa.Namespace, sa.Name)
+	mutex := getLock(lockKey)
 	mutex.Lock()
```

Comment:
Would it make sense to `TryLock` here and short circuit if it cannot be acquired? If a handler is already running for a given object, we won't have to do anything if it fails, as it will be re-enqueued anyway.

Reply:
The Go docs caution against TryLock

If it returns false we'd need to return an error to get a retry tho'?

Reply:
I don't think you would need a retry:
- goroutine 1 acquires lock
- goroutine 2 attempts to acquire lock
at this point, either goroutine 1 exits succesfully, and goroutine 2 does not need to run on the object, or goroutine 1 returns an error, which would cause the object to be reenqueued anyway. By always waiting on the lock, you run the risk of resource exhaustion if thousands of goroutines end up waiting for locks. I'm worried with the current approach that any handler that needs the secret to exist for the service account is always locking. Is there something we can do to only lock when the secret doesn't exist i.e we need to create it? i.e. if the get/list fails, acquire the lock, attempt another get/list (in case of race conditions), and then create? This way we don't have to lock every time we attempt to get the object, but rather only on creation?

Reply:
@bigkevmcd 

> If it returns false we'd need to return an error to get a retry tho'?

Correct. Any error returned by handlers ([apart from the special one](https://github.com/rancher/rancher/wiki/Tips-for-Writing-Rancher-Handlers#errors): `ErrIgnore`) will result in lasso requeuing the handler for later execution - [with a variable delay](https://github.com/rancher/rancher/wiki/Tips-for-Writing-Rancher-Handlers#enqueuing).

Reply:
@jakefhyde this method is not called inside of handlers exclusively - it's called for instance very close to the `main` of the `cattle-cluster-agent`, or in `Serve` methods that publish APIs, or inside of `kontainer-engine`.

All this places, AFAIU, will behave differently if we return an error that was not returned before just because of wrong timing.

In principle we could set a parameter to tell if it should block or not, although I am not sure it is worth the trouble.

For now we will be sending our customer a version of this PR more or less as it currently is and first of all observe if and how much it helps, and also benchmark if and how much lock contention really happens. Then I guess we can revisit this discussion for general inclusion for all customers.

Does that make sense from your perspective?

Reply:
It does but I still feel like we should only attempt to lock if the secret does not exist yet. For each handler there's only one point where they would need to create it, so I feel like it doesn't make any sense to lock if we aren't creating it.

Reply:
I'll take another look at locking only-on-creation.

Reply:
I think the initial issues we were seeing were due to the locking happening every update, so hopefully restricting to only on create will improve performance.

Reply:
@jakefhyde I think those were related to the old Lease mechanism that was put in.

Reply:
I think the problem with `lock-on-create` is that there's a race, we call `ServiceAccountSecret` which lists all secrets with a label matching the service account, then we iterate through that list, and check to see which one is "the one", and we attempt to delete the rest.

Then, if we didn't find a matching secret, we call `.Create()`.

While there's a small window here, I can easily reproduce the failure in tests, how close those are to the "real world" is debatable, but there's still a race-window.

Reply:
Maybe we can separate that into locking if there no suitable secret OR if there is more than one secret? From my perspective, we will only need to delete multiple secrets if more than one exists, and create if there is not a matching secret, but during normal flow there should only ever be 1 secret unless the existing one is deleted or a second is created, but I think it makes sense not to lock all the time. 

I won't block on it since CI is passing but it would be nice to do some scale testing for provisioning a few clusters at the same time to see how the performance behaves.